deps/src/singular.cpp

```diff
@@ -53,6 +75,46 @@ JLCXX_MODULE define_julia_module(jlcxx::Module & Singular)
     singular_define_ideals(Singular);
     singular_define_matrices(Singular);
 
+    Singular.method("call_interpreter", [](std::string input) {
+        // set callbacks
+        auto default_print = PrintS_callback;
+        auto default_error = WerrorS_callback;
+        auto default_warning = WarnS_callback;
+
+        PrintS_callback = PrintS_for_julia;
+        WerrorS_callback = WerrorS_for_julia;
+        WarnS_callback = WarningS_for_julia;
+
+        // cleanup return strings
+        singular_return.erase();
+        singular_error.erase();
+        singular_warning.erase();
+
+        // call interpreter
+        std::string  input_str = input + "\nreturn();";
+        const char * input_string = input_str.c_str();
+        char *       ost = (char *)omalloc0(strlen(input_string) + 1);
+        strcpy(ost, input_string);
+        bool err = iiAllStart(NULL, ost, BT_proc, 0);
+        inerror = 0;
+        errorreported = 0;
+        omFree(ost);
+
+        // Get output
+        jl_array_t * result = jl_alloc_array_1d(jl_array_any_type, 4);
+        jl_arrayset(result, err ? jl_true : jl_false, 0);
+        jl_arrayset(result, jl_cstr_to_string(singular_return.c_str()), 1);
+        jl_arrayset(result, jl_cstr_to_string(singular_error.c_str()), 2);
+        jl_arrayset(result, jl_cstr_to_string(singular_warning.c_str()), 3);
+
+        // reset hanlders
+        PrintS_callback = default_print;
+        WerrorS_callback = default_error;
+        WarnS_callback = default_warning;
```

Comment:
If you only temporarily change these callback in this function, why not replace them with lambdas? This way, you could also get rid of those ugly global strings.

Reply:
I do not want to reset those handlers completely, since I am not sure which errors they still export to the Julia terminal.

Lambdas are not possible here, as they would need to capture the strings, and one cannot have a function pointer from a lambda that captures. Relevant part from C++11 standard:
> The closure type for a lambda-expression with no lambda-capture has a public non-virtual non-explicit const conversion function to pointer to function having the same parameter and return types as the closure type’s function call operator. The value returned by this conversion function shall be the address of a function that, when invoked, has the same effect as invoking the closure type’s function call operator.

Reply:
I am not sure what you mean with "I do not want to reset those handlers completely", or perhaps in particular what "reset those handlers completely" means -- isn't that exactly what you are (and indeed, should be) doing here?

Reply:
I am setting them to my custom handlers while calling into the interpreter, and afterwards set them to their previous value. I do not want to overwrite them in `siInit` completely for the whole session.

Reply:
Of course they should only be changed temporarily here.